broker/broker-security-impl/src/main/java/org/codice/ddf/broker/security/impl/SubjectInjectorPlugin.java

```diff
@@ -0,0 +1,121 @@
+import org.codice.ddf.security.handler.api.UPAuthenticationToken;
+import org.codice.ddf.security.util.SAMLUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Element;
+
+public class SubjectInjectorPlugin implements BrokerMessageInterceptor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SubjectInjectorPlugin.class);
+
+  private static final Cache<String, Subject> SUBJECT_CACHE =
+      CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.MINUTES).build();
+
+  private static SecurityManager securityManager;
+
+  private static Set<String> configuredAddresses;
+
+  @Override
+  public void handleMessage(
+      ServerSession session,
+      Transaction tx,
+      Message message,
+      boolean direct,
+      boolean noAutoCreateQueue) {
+    if (!configuredAddresses.contains(message.getAddress())) {
+      return;
+    }
+    String subjectAsString = getStringSubjectFromSession(session);
+
+    message.putStringProperty("subject", subjectAsString);
+    if (message instanceof AMQPMessage) {
+      Map applicationPropertiesMap =
+          ((AMQPMessage) message).getProtonMessage().getApplicationProperties().getValue();
+      applicationPropertiesMap.put("subject", subjectAsString);
+      ((AMQPMessage) message)
+          .getProtonMessage()
+          .setApplicationProperties(new ApplicationProperties(applicationPropertiesMap));
+    }
+  }
+
+  public void setConfiguredAddresses(Set<String> addresses) {
+    configuredAddresses = new HashSet<>(addresses);
+  }
+
+  private Element getSubjectAsElement(ServerSession session) {
+    try {
+      return SUBJECT_CACHE
+          .get(session.getUsername(), () -> this.cacheAndReturnSubject(session))
+          .getPrincipals()
+          .oneByType(SecurityAssertion.class)
+          .getSecurityToken()
+          .getToken();
+    } catch (ExecutionException e) {
+      LOGGER.warn("Could not get Subject from token", e.getCause());
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  String getStringSubjectFromSession(ServerSession session) {
+    return SAMLUtils.getInstance().getSubjectAsStringNoSignature(getSubjectAsElement(session));
+  }
+
+  public void setSecurityManager(SecurityManager securityManager) {
+    SubjectInjectorPlugin.securityManager = securityManager;
+  }
+
+  public SecurityManager getSecurityManager() {
+    return securityManager;
+  }
+
+  public void clearCache() {
+    SUBJECT_CACHE.invalidateAll();
+  }
+
+  @VisibleForTesting
+  static Cache<String, Subject> getSubjectCache() {
+    return SUBJECT_CACHE;
+  }
+
+  @VisibleForTesting
+  Subject cacheAndReturnSubject(ServerSession session) throws SecurityServiceException {
+    UPAuthenticationToken usernamePasswordToken =
+        new UPAuthenticationToken(session.getUsername(), session.getPassword());
+    Subject subject = securityManager.getSubject(usernamePasswordToken);
+    SUBJECT_CACHE.put(session.getUsername(), subject);
```

Comment:
Why is the cache being modified here? This is your value loader function so Guava will update the cache with the value returned here.

Reply:
fixed